### PR TITLE
made certificate rotation idempotent

### PR DESCRIPTION
**Pull request by paynejacob**
#35343 

# Problem
If the planner runs the rotation plan more than once per generation  a server's certificate will be rotated multiple times

# Solution
track the generation on each node and only rotate certificates if the generation is different.

### CODE REVIEW

**Review by Oats87 — CHANGES_REQUESTED** [review]

File: pkg/provisioningv2/rke2/planner/certificaterotation.go [hunk 1]
```diff
@@ -86,11 +97,32 @@ func rotateCertificatesPlan(controlPlane *rkev1.RKEControlPlane, rotation *rkev1
...
+	if len(rotation.Services) > 0 {
+		for _, service := range rotation.Services {
+			args = append(args, "-s", service)
+		}
+	}
+
 	return plan.NodePlan{
+		Files: []plan.File{
+			{
+				Content: base64.StdEncoding.EncodeToString([]byte(idempotentRotateScript)),
```
Comment (annablender) [inline pkg/provisioningv2/rke2/planner/certificaterotation.go:118]: Question - this script is meant to be run programmatically right? Or manually? If it's the former(which I think it is), where do you pass the current and target gen in as args?

**Review by annablender — COMMENTED** [review]
Otherwise, LGTM syntax/logic wise though I'm not too familiar with cert stuff

File: pkg/provisioningv2/rke2/planner/certificaterotation.go (hunk 1 above)
Comment (jakefhyde) [reply pkg/provisioningv2/rke2/planner/certificaterotation.go:118]: The file is copied, but it is executed through the planner as part of the `OneTimeInstruction.Args`, so it's easy to miss: https://github.com/rancher/rancher/blob/b1b5da4362e5a4057c085cdae0ec644d929d3cfd/pkg/provisioningv2/rke2/planner/certificaterotation.go#L104

Comment (annablender) [reply pkg/provisioningv2/rke2/planner/certificaterotation.go:122]: I see, thanks!

**Review by jakefhyde — APPROVED** [review]

**Review by annablender — DISMISSED** [review]

**Review by Oats87 — APPROVED** [review]

**Review by jakefhyde — APPROVED** [review]

**Pull request merged**